### DISCUSSION TITLE
x-is component registration

### DISCUSSION THREAD

**Discussion by earthboundkid** [alpinejs/alpine#571] [4. General]
For Alpine 3, what if in HTML, you wrote `<div x-is="myThingie">…</div>` and then in JS, you called

```js
Alpine.component('myThingie', (el) => {
  return {
    open: false,
    init() { ... },
  };
});
```

Basically just sugar for `<div x-data="myThingie()" x-init="init()">…</div>` without needing to put myThingie on `window`.

**Comment by ryangjchandler** [comment]
This is already on the project roadmap, with a similar syntax but different directive to `x-is`.

**Reply by earthboundkid** [reply]
I think `x-is` is a good name for the directive because `is=` is a real thing in native custom elements. https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes#attr-is

What directive name were you planning to use?

**Reply by mblarsen** [reply]
Is there a public summarizing issue on the project road map on this issue and in general? @ryangjchandler

**Reply by ryangjchandler** [reply]
https://github.com/alpinejs/alpine/blob/master/README.md#v3-roadmap

There's no issue or discussion covering the idea. We can probably do that here, now that somebody else has made one. 

The original idea was covered in one of Caleb's live-streams, you might be able to find it in his YouTube, but we were torn between using `x-data` with the component name, or using a separate `x-component` directive IIRC.

**Reply by earthboundkid** [reply]
Another advantage of `x-is` is that it is shorter and easier to spell.

**Reply by ryangjchandler** [reply]
That's true, it's shorter but reading it sounds more like a comparison, than a declaration. I'd rather have a clearer, longer directive name that is easily understandable at a glance.

**Reply by HugoDF** [reply]
`is` seems to come from the Vue.js dynamic component syntax https://vuejs.org/v2/guide/components.html#Dynamic-Components but I don't know that we're leaning into "mostly Vue.js compatible" any more so we're likely to go with `x-component`.

**Reply by earthboundkid** [reply]
I don't know the history of Vue, but I imagine they were inspired by the HTML standard is https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/is. I don't think Alpine should use `is` because it's used in plain HTML and it's not a custom element system, so that leaves `x-is` instead. I guess you could confuse `x-is` with `x-if` or an `isX` variable if you weren't paying attention, but it's reasonably clear, and the word component is long and hard to spell.

**Reply by ryangjchandler** [reply]
But people would be using `Alpine.component()` anyway, so do you suggest we change that too?

**Reply by earthboundkid** [reply]
I don't like the `Alpine.component(...)` because component is a noun and not a verb. It should be `Alpine.defineComponent(...)` instead. Once you do that, it's not as big a deal that you use it with `x-is` instead of `x-component`. 

Here's a new idea though: the full attribute name is `x-component="..."` but the short name is `:is="..."`. Too crazy?

**Reply by SimoTod** [reply]
`:` is already a short version of `x-bind:`
It won't be easily possible because they clash and code should be deterministic. 
I'm not a fan of x-is for similar reasons, it's a verb and I feel it shouldn't be (but it's a personal opinion).
I would prefer `x-as` if we want a short syntax, although i don't think the word component is particularly long or hard to spell.
Out of curiosity, why is it so important that we use x-is? I assume the final decision will be on Caleb anyway but once defined, it's just a metter of remembering a key word.
Thanks for your feedback anyway.

**Reply by earthboundkid** [reply]
Good point about `:`. That won't work.

Using `is` is not particularly important. I've just been thinking about the custom elements API. Someone convinced the browsers to implement it, but AFAICT, it has almost zero uptake, in part because all it does is add a CSS reset (shadow DOM, semi-useful but niche) and lifecycle events (okay but could also be done in userspace). I'm just trying to think of whether there are any good ideas we can salvage from the mess. The `is=` attribute is about it. 😆
